libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFS20L.cpp

```diff
@@ -0,0 +1,150 @@
+        AP_RangeFinder_Params &_params,
+        AP_HAL::I2CDevice *dev)
+    : AP_RangeFinder_Backend(_state, _params)
+    , _dev(dev)
+{
+}
+
+AP_RangeFinder_Backend *AP_RangeFinder_Benewake_TFS20L::detect(
+        RangeFinder::RangeFinder_State &_state,
+        AP_RangeFinder_Params &_params,
+        AP_HAL::I2CDevice *dev)
+{
+    if (dev == nullptr) {
+        return nullptr;
+    }
+    AP_RangeFinder_Benewake_TFS20L *sensor = NEW_NOTHROW AP_RangeFinder_Benewake_TFS20L(_state, _params, dev);
+    if (sensor == nullptr || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+bool AP_RangeFinder_Benewake_TFS20L::init()
+{
+    uint8_t version_data[3];
+
+    {
+        WITH_SEMAPHORE(_dev->get_semaphore());
+        _dev->set_retries(3);
+
+        // Read firmware version to detect if sensor is present
+        if (!_dev->read_registers(TFS20L_VERSION_MAJOR, version_data, 3)) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "TFS20L: No response from sensor");
+            return false;
+        }
+
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "TFS20L: found fw version %u.%u.%u",
+                      version_data[0], version_data[1], version_data[2]);
+    }
+
+    // Register periodic callback to read sensor data
+    _dev->register_periodic_callback(50000,  // 20Hz sampling rate
+                                     FUNCTOR_BIND_MEMBER(&AP_RangeFinder_Benewake_TFS20L::timer, void));
+
+    return true;
+}
+
+void AP_RangeFinder_Benewake_TFS20L::update()
+{
+    WITH_SEMAPHORE(_sem);
+
+    if (accum.count > 0) {
+        state.distance_m = (accum.sum * 0.01f) / accum.count;
+        state.last_reading_ms = AP_HAL::millis();
+        accum.sum = 0;
+        accum.count = 0;
+        update_status();
+    } else if (AP_HAL::millis() - state.last_reading_ms > 200) {
+        set_status(RangeFinder::Status::NoData);
+    }
+}
+
+void AP_RangeFinder_Benewake_TFS20L::timer()
+{
+    uint8_t raw_data[4]; // Buffer for distance + strength data only
+
+    /*
+     * Read the first 4 registers to get distance and strength data
+     * TFS20L_DIST_LOW (0x00) through TFS20L_AMP_HIGH (0x03)
+     */
+    if (!_dev->read_registers(TFS20L_DIST_LOW, raw_data, sizeof(raw_data))) {
+        return;
+    }
+
+    // Combine bytes to form distance and strength (little-endian)
+    uint16_t distance_cm = (uint16_t(raw_data[1]) << 8) | raw_data[0];
+    uint16_t strength = (uint16_t(raw_data[3]) << 8) | raw_data[2];
+
+    // Validate reading and handle invalid cases
+    if (strength < MIN_STRENGTH || strength == 0xFFFF || 
+        distance_cm > MAX_DIST_CM || distance_cm < MIN_DIST_CM) {
+        /*
+         * When signal strength is too low or invalid, or distance is outside
+         * sensor's specified range, set distance to max + offset.
+         * This forces status to OutOfRangeHigh rather than NoData
+         */
+        distance_cm = max_distance()*100 + BENEWAKE_OUT_OF_RANGE_ADD_CM;
```

Comment:
This doesn't work - we can' accumulate a flag value like this - the flag nature of things doesn't allow for averaging :-)

Basically all readings have to be out of range high for you to declare out-of-range-high.

If this was modelled on another driver - which one? - because it needs fixing :-)